src/commonMain/kotlin/fr/acinq/lightning/crypto/KeyManager.kt

```diff
@@ -1,128 +1,103 @@
-         * - recompute your channel keys and spend your output
-         *
-         * @param fundingPubKey funding public key
-         * @return a BIP32 path
-         */
-        fun channelKeyPath(fundingPubKey: PublicKey): KeyPath {
-            val buffer = sha256(fundingPubKey.value)
-
-            val path = sequence {
-                var i = 0
-                while (true) {
-                    yield(Pack.int32BE(buffer, i).toUInt().toLong())
-                    i += 4
-                }
-            }
-            return KeyPath(path.take(8).toList())
+    data class ChannelKeys(
+        val fundingKeyPath: KeyPath,
+        val fundingPrivateKey: PrivateKey,
+        val paymentKey: PrivateKey,
+        val delayedPaymentKey: PrivateKey,
+        val htlcKey: PrivateKey,
+        val revocationKey: PrivateKey,
+        val shaSeed: ByteVector32
+    ) {
+        val fundingPubKey: PublicKey = fundingPrivateKey.publicKey()
+        val htlcBasepoint: PublicKey = htlcKey.publicKey()
+        val paymentBasepoint: PublicKey = paymentKey.publicKey()
+        val delayedPaymentBasepoint: PublicKey = delayedPaymentKey.publicKey()
+        val revocationBasepoint: PublicKey = revocationKey.publicKey()
+        val temporaryChannelId: ByteVector32 = (ByteVector(ByteArray(33) { 0 }) + revocationBasepoint.value).sha256()
+        fun commitmentPoint(index: Long): PublicKey = Bolt3Derivation.perCommitPoint(shaSeed, index)
+        fun commitmentSecret(index: Long): PrivateKey = Bolt3Derivation.perCommitSecret(shaSeed, index)
+    }
+
+    data class Bip84OnChainKeys(
+        private val chain: NodeParams.Chain,
+        val account: Long,
+        val xpriv: DeterministicWallet.ExtendedPrivateKey
+    ) {
+        constructor(chain: NodeParams.Chain, master: DeterministicWallet.ExtendedPrivateKey, account: Long) : this(
+            chain, account,
+            xpriv = DeterministicWallet.derivePrivateKey(master, bip84BasePath(chain) I hardened(account))
+        )
+
+        fun privateKey(addressIndex: Long): PrivateKey {
+            return DeterministicWallet.derivePrivateKey(xpriv, KeyPath.empty I 0 I addressIndex).privateKey
+        }
+
+        fun pubkeyScript(addressIndex: Long): ByteVector {
+            val priv = privateKey(addressIndex)
+            val pub = priv.publicKey()
+            val script = Script.pay2wpkh(pub)
+            return Script.write(script).toByteVector()
+        }
+
+        fun address(addressIndex: Long): String {
+            return Bitcoin.computeP2WpkhAddress(privateKey(addressIndex).publicKey(), chain.chainHash)
         }
 
-        fun channelKeyPath(fundingPubKey: ExtendedPublicKey): KeyPath = channelKeyPath(fundingPubKey.publicKey)
+        val xpub: String = DeterministicWallet.encode(
+            input = DeterministicWallet.publicKey(DeterministicWallet.derivePrivateKey(xpriv, KeyPath.empty I hardened(account))),
```

Comment:
Shouldn't this just be:

```suggestion
            input = DeterministicWallet.publicKey(xpriv),
```

Reply:
Nice catch! Fixed and added test in aa5dc819.